mailcheck.1
.TH MAILCHECK 1 "2 July 2005"

.SH NAME
mailcheck \- Check multiple mailboxes and/or Maildirs for new mail

.SH SYNOPSIS
\fBmailcheck\fP [-sh] [-f rcfile]

.SH DESCRIPTION
\fBmailcheck\fP is a simple, configurable tool that allows multiple
mailboxes to be checked for the existence of mail.  For local mail, it
supports both the traditional mbox format and the newer Maildir format.
.PP
Typically, one would invoke \fBmailcheck\fP in /etc/profile or a
user-specific login script.  E-mail junkies may also find it useful to
invoke \fBmailcheck\fP occasionally to check for new mail in alternate
mailboxes.
.PP
The author uses \fBmailcheck\fP to keep track of messages arriving in
mailboxes corresponding to several mailing lists he subscribes to.

.SH OPTIONS
.TP
\fB\-s\fP
Print "no mail" summary or total number of new/unread messages.
Do not print values for every mailbox.
.TP
\fB\-f\fP
Specify alternative rc file location.  If provided, default locations (see
\fBFILES\fP) are not checked.
.TP
\fB\-h\fP
Print short usage information.

.SH CONFIGURATION
Configuring \fBmailcheck\fP is simple.  Upon startup, \fBmailcheck\fP looks
for a file called \fB.mailcheckrc\fP in the user's home directory.  If that
file does not exist, the default configuration file \fB/etc/mailcheckrc\fP
is used instead.
.PP
Lines beginning with a hash sign (\fB#\fP) are treated as comments and will
not be processed.  All other lines are treated as pathnames to mailbox files
or Maildir directories.
.PP
Environment variables in the format \fB$(NAME)\fP will be expanded inline.
For example: 
.TP
\fB/var/spool/mail/$(USER)\fP 
Will check the user's mailbox in \fB/var/spool/mail\fP.
.TP
\fB$(HOME)/Mailbox\fP
Will check the default mailbox used by qmail installations.
.PP

.SH FILES
.TP
.B /etc/mailcheckrc
This is the site-default mailcheck configuration file.  It should be
edited by the system administrator to meet the needs of most users on
the system.
.TP
.B ~/.mailcheckrc
This is the user-specific mailcheck configuration file.  If it exists
for a particular user, the site-default configuration file will not
be used.

.SH COPYRIGHT
Copyright (C) 1996, 1997, 1998, 2001, Jefferson E. Noxon.
.PP
Portions Copyright (C) 1996, Free Software Foundation, Inc.
.PP
Portions Copyright (C) 1996, Gordon Matzigkeit.
.PP
Portions Copyright (C) 1998, Trent Piepho.
.PP
Other copyrights may apply.
.PP
This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.
.PP
On Debian GNU/Linux see /usr/share/common-licenses/GPL

.SH AUTHOR
Mailcheck was written for Debian GNU/Linux by Jefferson E. Noxon
<jeff@planetfall.com>.

.SH ACKNOWLEDGEMENTS
POP3 and IMAP support was added by Rob Funk <rfunk@funknet.net>.
.PP
Several enhancements by Tomas Hoger <thoger@pobox.sk>.

.SH BUGS
It is probably not a good idea to store passwords in a .netrc file.
.PP
No SSL/TLS support for POP3 and IMAP.

.SH SEE ALSO
netrc(5), mbox(5), maildir(5), login(1), fetchmail(1)



debian/changelog
mailcheck (1.91.2-2-ke5) unstable; urgency=low

  * Shorter output

 -- Kamil Essekkat <kamil@essekkat.pl>  Mon, 03 Mar 2014 15:55:38 +0100

mailcheck (1.91.2-2-ke4) unstable; urgency=low

  * Be quiet if no mail is found

 -- Kamil Essekkat <kamil@essekkat.pl>  Mon, 03 Mar 2014 15:28:36 +0100

mailcheck (1.91.2-2-ke3) unstable; urgency=low

  * Summary mode enhanced and exclusive with normal mode.

 -- Kamil Essekkat <kamil@essekkat.pl>  Mon, 03 Mar 2014 14:43:25 +0100

mailcheck (1.91.2-2-ke2) unstable; urgency=low

  * Usage information updated.

 -- Kamil Essekkat <kamil@essekkat.pl>  Tue, 08 Jan 2013 10:06:39 +0100

mailcheck (1.91.2-2-ke1) unstable; urgency=low

  * Non-maintainer upload.
  * Removed pop3 and imap support.

 -- Kamil Essekkat <kamil@essekkat.pl>  Tue, 18 Dec 2012 14:38:14 +0100

mailcheck (1.91.2-2) unstable; urgency=low

  * Added advanced_counting patch from Tomas Hoger <thoger@pobox.sk>.
  * Created new patch to removing stripping in Makefile. Closes: #437550.
  * Added Tomas Hoger <thoger@pobox.sk> to debian/copyright.
  * Changed distclean invocation in debian/rules and cleaned up some cruft.
  * New e-mail address for maintainer. Changed throughout debian/* files.

 -- Kevin Coyner <kcoyner@debian.org>  Tue, 14 Aug 2007 12:53:50 -0400

mailcheck (1.91.2-1) unstable; urgency=low

  * New maintainer. Closes: #404469.
  * debian/control:
    + Updated debhelper to version 5.0
    + Updated Standards-Version to 3.7.2
    + Added dpatch as a Build-Depends
    + Minor grammatical correction in long description
    + Added homepage link.
  * Reorganized debian/copyright.
  * debian/rules:
    + Added appropriate language for using dpatch
    + Cleaned up unneeded comments, obsolete dh_installmanpages
  * Added debian/compat file.
  * Removed debian/conffiles, since dh_installdeb adds any conffiles
    automatically
  * Applied patch that corrects problems with @'s in username.  Thanks to
    Martin Wickman, who provided this patch more than 2 years ago.
    Closes: #272749.
  * Acknowledge NMU from and thanks extended to Daniel Schepler.
    Closes: #190484.
  * Uploaded original source files to SourceForge (mailcheck) in order to
    convert this package from a Debian native to a normal package.
  * Added debian/watch file.

 -- Kevin Coyner <kevin@rustybear.com>  Sat, 3 Mar 2007 06:26:05 -0500

mailcheck (1.91.1) unstable; urgency=low

  * NMU.
  * Add Build-Depends on debhelper.  Closes: #190484.

 -- Daniel Schepler <schepler@debian.org>  Sun,  3 Aug 2003 00:55:16 -0700

mailcheck (1.91) unstable; urgency=low

  * Applied patch to add POP3 & IMAP support (closes: Bug#96911)
  * Don't report negative mail messages on invalid maildirs (closes: Bug#97163)
  * Updated to debian-policy 3.5.4.0
  * "copyright" file pointed to wrong directory on Debian systems

 -- Jefferson E. Noxon <jeff@debian.org>  Fri, 11 May 2001 17:04:01 -0500

mailcheck (1.0) unstable; urgency=low

  * Updated to debian-policy 3.0.1.1
  * Changed maintainer address

 -- Jefferson E. Noxon <jeff@debian.org>  Wed, 15 Sep 1999 14:36:31 -0500

mailcheck (0.98) unstable; urgency=low

  * First public release
  * Converted to debhelper

 -- Jefferson E. Noxon <jeff@planetfall.com>  Wed, 26 Aug 1998 14:27:49 -0500

mailcheck (0.97) unstable; urgency=low

  * Updated to libc6

 -- Jefferson E. Noxon <jeff@planetfall.com>  Thu,  5 Mar 1998 14:15:51 -0600

mailcheck (0.91) unstable; urgency=low

  * Added rc file support

 -- Jefferson E. Noxon <jeff@planetfall.com>  Fri, 27 Dec 1996 15:45:26 -0600

mailcheck (0.9) unstable; urgency=low

  * Initial Release.

 -- Jefferson E. Noxon <jeff@planetfall.com>  Fri, 27 Dec 1996 13:45:35 -0600


mailcheck.c
/* mailcheck.c
 *
 * Copyright 1996, 1997, 1998, 2001 Jefferson E. Noxon <jeff@planetfall.com>
 *           2001 Rob Funk <rfunk@funknet.net>
 *           2003, 2005 Tomas Hoger <thoger@pobox.sk>
 *
 * This file may be copied under the terms of the GNU Public License
 * version 2, incorporated herein by reference.
 */

/* Command line parameters:
 * -s: print "no mail" summary if needed
 * -f: specify alternative rc file location
 * -h: print usage
 */

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <dirent.h>
#include <unistd.h>
#include <fcntl.h>
#include <ctype.h>

#define BUF_SIZE (2048)


/* Global variables */
char *Homedir;			/* Home directory pathname */
char info[4000];
unsigned short global_new = 0;
unsigned short global_unread = 0;
struct {
  unsigned short show_summary;		/* see '-s' option */
  char *rcfile_path;			/* see '-f' option */
} Options= {0, NULL};
  

/* Print usage information. */
void 
print_usage(void)
{
  printf("Usage: mailcheck [-hs] [-f rcfile]\n"
         "\n"
         "Options:\n"
         "  -s  - show \"no mail\" summary, if no new mail was found\n"
         "  -f  - specify alternative rcfile location\n"
         "  -h  - show this help screen\n"
         "\n");
}

/* Open an rc file.  Exit with error message, if attempt to open rcfile failed.
 * Otherwise, return valid FILE* .
 */
FILE *
open_rcfile (void)
{
  char namebuf[256];
  FILE *rcfile;

  /* if rcfile path was provided, do not try default locations */
  if (Options.rcfile_path != NULL) {
    if ((rcfile= fopen(Options.rcfile_path, "r")) == NULL) {
      fprintf(stderr, "error: couldn't open rcfile '%s'\n",
	  Options.rcfile_path);
      exit(1);
    }
  }
  else {
    snprintf(namebuf, sizeof (namebuf), "%s/.mailcheckrc", Homedir);
    
    if ((rcfile= fopen(namebuf, "r")) == NULL) {
      if ((rcfile= fopen("/etc/mailcheckrc", "r")) == NULL) {
        fprintf (stderr, "mailcheck: couldn't open /etc/mailcheckrc "
            "nor %s/.mailcheckrc\n", Homedir);
	exit(1);
      }
    }
  }
  
  return rcfile;
}


/* Expand environment variables.  Input buffer must be long enough
 * to hold output.  Nested $()'s don't work.
 */
char *
expand_envstr (char *path)
{
  char *srcptr, *envptr, *tmpptr, *dup, *envname;
  int len;

  srcptr = strstr (path, "$(");
  if (!srcptr)
    return path;

  len = strlen (path);
  if (!len)
    return path;

  dup = malloc (len + 1);
  if (!dup)
    return path;

  strcpy (dup, path);
  envptr = strstr (dup, "$(") + 2;
  tmpptr = strchr (envptr, ')');
  if (!tmpptr)
    {
      free (dup);
      return path;
    }

  *tmpptr = 0;
  *srcptr = 0;
  envname = getenv (envptr);
  if (envname)
    strcat (srcptr, envname);
  strcat (srcptr, tmpptr + 1);

  free (dup);
  return expand_envstr (path);
}

/* Should entry in maildir be ignored? */
inline int
ignore_maildir_entry(const char *dir, const struct dirent *entry) {
  char fname[BUF_SIZE];
  struct stat filestat;
  
  /* *all* dotfiles should be ignored in maildir, not only . and .. ! */
  if (entry->d_name[0] == '.')
    return 1;

  /* also count only regular files
   * use dirent's d_type if possible, otherwise stat file (which is much
   * slower) */
#ifdef _DIRENT_HAVE_D_TYPE
  if (entry->d_type != DT_UNKNOWN) {
    if (entry->d_type != DT_REG)
      return 1;
  }
  else {
#endif /* _DIRENT_HAVE_D_TYPE */
    snprintf(fname, sizeof(fname), "%s/%s", dir, entry->d_name);
    fname[sizeof(fname) - 1]= '\0';

    if (stat(fname, &filestat) != 0) {
      fprintf(stderr, "mailcheck: failed to stat file: %s\n", fname);
      return 1;
    }
    
    if (!S_ISREG(filestat.st_mode))
      return 1;
#ifdef _DIRENT_HAVE_D_TYPE
  }
#endif /* _DIRENT_HAVE_D_TYPE */

  return 0;
}

/* Count files in subdir of maildir (new/cur/tmp). */
int
count_entries (char *path)
{
  DIR *mdir;
  struct dirent *entry;
  int count = 0;

  if ((mdir= opendir(path)) == NULL)
    return -1;

  while ((entry = readdir (mdir))) {
    if (ignore_maildir_entry(path, entry)) 
      continue;
    
    count++;
  }

  closedir(mdir);

  return count;
}

/* Count mails in unix mbox. */
int
check_mbox(const char *path, int *new, int *read, int *unread)
{
  char linebuf[BUF_SIZE];
  FILE *mbox;
  int linelen;
  unsigned short in_header= 0;	/* do we parse mail header or mail body? */

  if ((mbox= fopen(path, "r")) == NULL) {
    fprintf(stderr, "mailcheck: unable to open mbox %s\n", path);
    return -1;
  }

  *new= 0;
  *read= 0;
  *unread= 0;
  
  while (fgets(linebuf, sizeof(linebuf), mbox)) {
    if (!in_header) {
      if (strncmp(linebuf, "From ", 5) == 0) { /* 5 == strlen("From ") */
	in_header= 1;
	(*new)++;
      }
    }
    else {
      if (linebuf[0] == '\n') {
	in_header= 0;
      }
      else if (strncmp(linebuf, "Status: ", 8) == 0) { /* 8 == strlen("Status: ") */
	linelen= strlen(linebuf);
	
	if (linelen >= 10  && 
	    ((linebuf[8] == 'R'  &&  linebuf[9] == 'O')  ||
	     (linebuf[8] == 'O'  &&  linebuf[9] == 'R'))) {
	  (*new)--;
	  (*read)++;
	}
	else if (linelen >= 9  &&  linebuf[8] == 'O') {
	  (*new)--;
	  (*unread)++;
	}
      }
    }
  }

  fclose(mbox);

  return 0;
}

/* Count mails in maildir.  Newer, more sophisticated, but also more time
 * consuming version. */
int
check_maildir(const char *path, int *new, int *read, int *unread) {
  char dir[BUF_SIZE];
  DIR *mdir;
  struct dirent *entry;
  char *pos;

  /* new mail - standard way */
  snprintf(dir, sizeof(dir), "%s/new", path);
  *new= count_entries(dir);
  if (*new == -1)
    return -1;
  
  /* older mail - check also mail status */
  snprintf(dir, sizeof(dir), "%s/cur", path);
  if ((mdir= opendir(dir)) == NULL)
    return -1;

  *read= 0;
  *unread= 0;
  while ((entry = readdir (mdir))) {
    if (ignore_maildir_entry(dir, entry)) 
      continue;

    if ((pos= strchr(entry->d_name, ':')) == NULL) {
      (*unread)++;
    } 
    else if (*(pos + 1) != '2') {
      fprintf(stderr, "mailcheck: ooops, unsupported experimental info "
	  "semantics on %s/%s\n", dir, entry->d_name);
      continue;
    }
    else if (strchr(pos, 'S') == NULL) {
      					/* search for seen ('S') flag */
      (*unread)++;
    }
    else {
      (*read)++;
    }
  }

  closedir(mdir);
  
  return 0;
}


/* Check for mail in given mail path (could be mbox, maildir). */
void
check_for_mail (char *tmppath)
{
  struct stat st;
  char *mailpath;

  /* expand environment variables in path specifier */
  mailpath= expand_envstr (tmppath);
  
  if (!stat (mailpath, &st)) {
    /* Is it regular file? (if yes, it should be mailbox ;) */
    if (S_ISREG (st.st_mode)) {
        int new, read, unread;
        if (check_mbox(mailpath, &new, &read, &unread) == -1)
          return;
	if (new > 0 || unread > 0) {
          sprintf(info, "%s: %d new and %d unread message(s)\n", mailpath, new, unread);
          global_new += new;
          global_unread += unread;
        }
    }
    /* Is it directory? (if yes, it should be maildir ;) */
    /* for maildir specification, see: http://cr.yp.to/proto/maildir.html */
    else if (S_ISDIR (st.st_mode)) {
	int new, read, unread;

	if (check_maildir(mailpath, &new, &read, &unread) == -1) {
	  fprintf (stderr, 
	      "mailcheck: %s is not a valid maildir -- skipping.\n", mailpath);
	  return;
	}
	if (new > 0 || unread > 0) {
          sprintf(info, "%s: %d new and %d unread message(s)\n", mailpath, new, unread);
          global_new += new;
          global_unread += unread;
        }
    } else {
      fprintf(stderr, "mailcheck: error, %s is not mbox or maildir\n",
	  mailpath);
      }
  }
  else {
    fprintf(stderr, "mailcheck: invalid line '%s' in rc-file\n", mailpath);
  }
}

/* Process command-line options */
void
process_options (int argc, char *argv[])
{
  int opt;
  
  while ((opt= getopt(argc, argv, "hsf:")) != -1)
    {
      switch (opt)
	{
	case 'h':
	  print_usage();
	  exit(0);
	  break;
	case 's':
	  Options.show_summary= 1;
	  break;
	case 'f':
	  Options.rcfile_path= optarg;
	  break;
	}
    }
}

/* main */
int
main (int argc, char *argv[])
{
  char buf[1024], *ptr;
  FILE *rcfile;

  ptr= getenv ("HOME");
  if (!ptr) {
    fprintf (stderr, "mailcheck: couldn't read environment variable HOME.\n");
    return 1;
  } else {
    Homedir= strdup(ptr);
  }

  process_options (argc, argv);

  rcfile= open_rcfile();

  while (fgets (buf, sizeof (buf), rcfile))
    {
      /* eliminate newline */
      ptr = strchr (buf, '\n');
      if (ptr)
	*ptr = '\0';

      /* If it's not a blank line or comment, look for mail in it */
      if (strlen (buf) && (*buf != '#'))
	check_for_mail (buf);
    }

  if (Options.show_summary && (global_new || global_unread)) {
    char info[30];
    if (global_new) { sprintf(info, "%d new", global_new); }
    if (global_unread) { sprintf(info, "%d unread", global_unread); }
    printf("%s\n", info);
  } else {
    printf(info);
  }
  
  fclose(rcfile);
  free(Homedir);

  return 0;
}

/* vim:set ts=8 sw=2: */
